File: runtime/src/api/specClient.ts
import { ComponentTree } from '../types';

/** Preview params captured once at module load (before SPA navigation strips them). */
const _previewQs: string = (() => {
  const pageParams = new URLSearchParams(window.location.search);
  const qs = new URLSearchParams();
  for (const key of ['specId', 'versionId', 'compId']) {
    const val = pageParams.get(key);
    if (val) qs.set(key, val);
  }
  return qs.toString();
})();

export async function fetchSpec(): Promise<ComponentTree> {
  const suffix = _previewQs ? `?${_previewQs}` : '';
  const res = await fetch(`/_renderer/spec${suffix}`);
  if (!res.ok) {
    throw new Error(`Failed to fetch spec: ${res.status}`);
  }
  return res.json();
}

File: flutter_runtime/lib/api/spec_client.dart
import 'package:dio/dio.dart';
import '../models/component_tree.dart';

/// Preview params captured once at load time (before GoRouter navigation strips them).
final Map<String, String> _previewParams = () {
  final params = Uri.base.queryParameters;
  final result = <String, String>{};
  for (final key in ['specId', 'versionId', 'compId']) {
    final val = params[key];
    if (val != null && val.isNotEmpty) {
      result[key] = val;
    }
  }
  return result;
}();

class SpecClient {
  final Dio _dio;

  SpecClient({String? baseUrl})
      : _dio = Dio(BaseOptions(
          baseUrl: baseUrl ?? 'http://localhost:3000',
          connectTimeout: const Duration(seconds: 5),
          receiveTimeout: const Duration(seconds: 10),
        ));

  Future<ComponentTree> fetchSpec() async {
    final response = await _dio.get(
      '/_renderer/spec',
      queryParameters: _previewParams.isNotEmpty ? _previewParams : null,
    );
    return ComponentTree.fromJson(response.data as Map<String, dynamic>);
  }
}
